File: robot/robot/robot_node.py
import rclpy
from rclpy.node import Node

import math
from scipy.spatial.transform import Rotation

from geometry_msgs.msg import PoseStamped, Point, Quaternion, Twist


class Robot(Node):

    v = 0.0
    w = 0.0
    x = 0.0
    y = 0.0
    theta = 0.0
    time = 0.0
    lasttime = 0.0

    def __init__(self):
        super().__init__('robot')
        self.lasttime = self.get_clock().now()
        self.subscription = self.create_subscription(
            Twist,
            'cmd_vel',
            self.listener_callback,
            10)
        self.subscription

        self.publisher_ = self.create_publisher(PoseStamped, 'display', 10)
        timer_period = 0.01
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.i = 0

    def listener_callback(self, msg):
        self.v = msg.linear.x
        self.w = msg.angular.z


    def matadd(self, A, B):
        return [[A[i][j] + B[i][j]  for j in range(len(A[0]))] for i in range(len(A))]

    def matmul(self, A, B):
        return [[sum(a*b for a,b in zip(A_row, B_col)) for B_col in zip(*B)] for A_row in A]
        


    def timer_callback(self):
        self.time = self.get_clock().now()
        dt = self.time - self.lasttime
        dt_nano = dt.nanoseconds / 1e9

        A = [[1.0, 0.0, 0.0],
             [0.0, 1.0, 0.0],
             [0.0, 0.0, 1.0]]
        
        xt = [[self.x],
              [self.y],
              [self.theta]]
        
        B = [[math.cos(self.theta) * dt_nano, 0,0],
             [math.sin(self.theta) * dt_nano, 0,0],
             [0.0, dt_nano]]

        ut = [[self.v],
              [self.w]]
        
        Ax = self.matmul(A, xt)
        Bu = self.matmul(B, ut)

        self.lasttime = self.time

        result = self.matadd(Ax, Bu)

        self.x = result[0][0]
        self.y = result[1][0]
        self.theta = result[2][0]


        msg = PoseStamped()
        msg_point = Point()
        msg_quater = Quaternion()

        msg_point.x = self.x
        msg_point.y = self.y
        msg_point.z = 0.0

        rot = Rotation.from_euler('xyz', [0, 0, self.theta], degrees=False)
        rot_quat = rot.as_quat()

        msg_quater.x = rot_quat[0]
        msg_quater.y = rot_quat[1]
        msg_quater.z = rot_quat[2]
        msg_quater.w = rot_quat[3]

        
        msg.header.frame_id = "map"
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.pose.position = msg_point
        msg.pose.orientation = msg_quater


        self.get_logger().info(f"Result: {msg}, theta: {self.theta}")


        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)

    robot = Robot()

    rclpy.spin(robot)

    robot.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cpp_robot/src/robot_node.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <cmath>

#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include "geometry_msgs/msg/pose_stamped.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "geometry_msgs/msg/quaternion.hpp"

#include "tf2/LinearMath/Quaternion.h"

using namespace std::chrono_literals;
using std::placeholders::_1;


class MinimalPublisher : public rclcpp::Node
{
public:
    MinimalPublisher() : Node("minimal_publisher"), count_(0)
    {
        this->lasttime = this->get_clock()->now().nanoseconds();
        this->subscription_ = this->create_subscription<geometry_msgs::msg::Twist>(
        "topic", 10, std::bind(&MinimalPublisher::listener_callback, this, _1));

        publisher_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("display", 10);
        timer_ = this->create_wall_timer(
        500ms, std::bind(&MinimalPublisher::timer_callback, this));
    }

private:
    void multiplyBu(float firstMatrix[][2], float secondMatrix[][1], float mult[][1], int rowFirst, int columnFirst, int rowSecond, int columnSecond)
    {
        int i, j, k;

        // Initializing elements of matrix mult to 0.
        for(i = 0; i < rowFirst; ++i)
        {
            for(j = 0; j < columnSecond; ++j)
            {
                mult[i][j] = 0;
            }
        }

        // Multiplying matrix firstMatrix and secondMatrix and storing in array mult.
        for(i = 0; i < rowFirst; ++i)
        {
            for(j = 0; j < columnSecond; ++j)
            {
                for(k=0; k<columnFirst; ++k)
                {
                    mult[i][j] += firstMatrix[i][k] * secondMatrix[k][j];
                }
            }
        }
    }

    void multiplyAx(float firstMatrix[][3], float secondMatrix[][1], float mult[][1], int rowFirst, int columnFirst, int rowSecond, int columnSecond)
    {
        int i, j, k;

        // Initializing elements of matrix mult to 0.
        for(i = 0; i < rowFirst; ++i)
        {
            for(j = 0; j < columnSecond; ++j)
            {
                mult[i][j] = 0;
            }
        }

        // Multiplying matrix firstMatrix and secondMatrix and storing in array mult.
        for(i = 0; i < rowFirst; ++i)
        {
            for(j = 0; j < columnSecond; ++j)
            {
                for(k=0; k<columnFirst; ++k)
                {
                    mult[i][j] += firstMatrix[i][k] * secondMatrix[k][j];
                }
            }
        }
    }

    void addMatrices(float a[3][1], float b[3][1], float sum[3][1], int r, int c)
    {
        int i, j;

        for(i = 0; i < r; ++i)
            for(j = 0; j < c; ++j)
                sum[i][j] = a[i][j] + b[i][j];
        }

    void timer_callback()
    {
        this->time = this->get_clock()->now().nanoseconds();
        float dt = this->time - this->lasttime;

        float A[3][3] = {{1.0, 0.0, 0.0},
                         {0.0, 1.0, 0.0},
                         {0.0, 0.0, 1.0}};

        float xt[3][1] = {{this->x},
                          {this->y},
                          {this->theta}};

        float B[3][2] = {{cosf(this->theta) * dt, 0.0},
                         {sinf(this->theta) * dt, 0.0},
                         {0.0, dt}};

        float ut[2][1] = {{this->v},
                          {this->w}};

        float Ax[3][1];
        float Bu[3][1];
        this->multiplyAx(A, xt, Ax, 3, 3, 3, 1);
        this->multiplyBu(B, ut, Bu, 3, 3, 3, 1);

        this->lasttime = this->time;

        float result[3][1];
        this->addMatrices(Ax, Bu, result, 3, 1);

        auto msg = geometry_msgs::msg::PoseStamped();
        auto msg_point = geometry_msgs::msg::Point();
        auto msg_quat = geometry_msgs::msg::Quaternion();

        tf2::Quaternion quat;
        quat.setRPY(0.0, 0.0, this->theta);
        quat = quat.normalize();

        msg_quat.set__x(quat.getX());
        msg_quat.set__y(quat.getY());
        msg_quat.set__z(quat.getZ());
        msg_quat.set__w(quat.getW());

        msg_point.set__x(this->x);
        msg_point.set__y(this->y);
        msg_point.set__z(0.0);

        msg.header.frame_id = "map";
        msg.header.stamp = this->now();
        msg.pose.position = msg_point;
        msg.pose.orientation = msg_quat;

        RCLCPP_INFO(this->get_logger(), "I heard: '%lf'", msg.pose.position.x);

        publisher_->publish(msg);
    }

    void listener_callback(const geometry_msgs::msg::Twist::SharedPtr msg)
    {
        this->v = msg->linear.x;
        this->w = msg->angular.z;
    }

    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr publisher_;
    rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr subscription_;
    size_t count_;

    float v, w, x, y, theta, time, lasttime;

};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<MinimalPublisher>());
    rclcpp::shutdown();
    return 0;
}